test/org/javarosa/xpath/expr/XPathPathExprCurrentTest.java

```diff
@@ -0,0 +1,100 @@
+import org.javarosa.core.model.data.StringData;
+import org.javarosa.core.model.data.helper.Selection;
+import org.javarosa.core.model.instance.InstanceInitializationFactory;
+import org.javarosa.core.test.FormParseInit;
+import org.javarosa.form.api.FormEntryController;
+import org.javarosa.form.api.FormEntryModel;
+import org.javarosa.form.api.FormEntryPrompt;
+import org.junit.Before;
+import org.junit.Test;
+
+import static junit.framework.TestCase.assertEquals;
+import static org.javarosa.test.utils.ResourcePathHelper.r;
+
+public class XPathPathExprCurrentTest {
+    private FormDef formDef;
+    private FormEntryController formEntryController;
+
+    @Before
+    public void setUp() {
+        FormParseInit fpi = new FormParseInit(r("relative-current-ref.xml"));
+        formDef = fpi.getFormDef();
+        formDef.initialize(true, new InstanceInitializationFactory());
+        FormEntryModel formEntryModel = new FormEntryModel(formDef);
+        formEntryController = new FormEntryController(formEntryModel);
+    }
+
+    /**
+     * current() in a calculate should refer to the node it is in (in this case, /data/my_group/name_relative).
+     * This means that to refer to a sibling node, the path should be current()/../<name of sibling node>. This is
+     * verified by changing the value of the node that the calculate is supposed to refer to
+     * (/data/my_group/name) and seeing that the dependent calculate is updated accordingly.
+     */
+    @Test
+    public void current_in_calculate_should_refer_to_node() {
+        formEntryController.jumpToFirstQuestionWithName("name");
+        StringData nameValue = new StringData("Bob");
+        formEntryController.answerQuestion(nameValue, true);
+
+        StringData relativeName = (StringData) formDef.getFirstDescendantWithName("name").getValue();
+        assertEquals(nameValue.getValue(), relativeName.getValue());
+    }
+
+    /**
+     * current() in a choice filter should refer to the select node the choice filter is called from, NOT the expression
+     * it is in. See https://developer.mozilla.org/en-US/docs/Web/XPath/Functions/current -- this is the difference
+     * between current() and .
+     *
+     * The behavior of current() in a choice filter is verified by selecting a value for a first, static select and then
+     * using that value to filter a second, dynamic select.
+     */
+    @Test
+    public void current_in_choice_filter_should_refer_to_node() {
+        formEntryController.jumpToFirstQuestionWithName("fruit");
+        Selection fruitSelection = new Selection(1);
+        fruitSelection.attachChoice(formEntryController.getModel().getQuestionPrompt().getQuestion());
+        formEntryController.answerQuestion(new SelectOneData(fruitSelection), true);
+
+        SelectOneData fruitSelectionValue = (SelectOneData) formDef.getFirstDescendantWithName("fruit").getValue();
+        assertEquals("blueberry", fruitSelectionValue.getDisplayText());
```

Comment:
Is this just for confirmation?

Reply:
I built the answer by selecting an index of 1. This gives me some certainty that I have actually picked the value I had intended to so that if the assertions below fail I know for sure it's not because of an issue with the static select or the form or something like that. Tests that are based on reading form XML can break catastrophically because of a small form change so I like to have breadcrumbs there ensuring the different steps actually did what they were supposed to.

Reply:
Then I'd like to suggest putting a comment to explain that the assertion is not part of the test.

Reply:
Sure. Something weird happened but I think you saw my comment that used to be here?

> I built the answer by selecting an index of 1. This gives me some certainty that I have actually picked the value I had intended to so that if the assertions below fail I know for sure it's not because of an issue with the static select or the form or something like that. Tests that are based on reading form XML can break catastrophically because of a small form change so I like to have breadcrumbs there ensuring the different steps actually did what they were supposed to.

Reply:
Yup